cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i

```diff
@@ -0,0 +1,304 @@
+            {
+                if (classWeights[j] > maxWeight)
+                {
+                    maxWeight      = classWeights[j];
+                    maxWeightClass = j;
+                }
+            }
+            testLabel[i] = maxWeightClass;
+        }
+        return services::Status();
+    }
+
+    services::Status computeDistances(const NumericTable * trainTable, const NumericTable * testTable)
+    {
+        daal::SafeStatus s;
+
+        const size_t nDims  = trainTable->getNumberOfColumns();
+        const size_t nTrain = trainTable->getNumberOfRows();
+        const size_t nTest  = testTable->getNumberOfRows();
+
+        _distances = static_cast<FPType *>(services::internal::service_malloc<FPType, cpu>(nTrain * nTest * sizeof(FPType)));
+
+        daal::algorithms::internal::EuclideanDistances<FPType, cpu> euclDist(*testTable, *trainTable);
+        euclDist.init();
+
+        const size_t blockSize    = 128;
+        const size_t nOuterBlocks = nTest / blockSize + !!(nTest % blockSize);
+        const size_t nInnerBlocks = nTrain / blockSize + !!(nTrain % blockSize);
+
+        daal::threader_for(nOuterBlocks, nOuterBlocks, [&](size_t outerBlock) {
+            const size_t outerStart = outerBlock * blockSize;
+            const size_t outerEnd   = outerBlock + 1 == nOuterBlocks ? nTest : outerStart + blockSize;
+            const size_t outerSize  = outerEnd - outerStart;
+
+            daal::threader_for(nInnerBlocks, nInnerBlocks, [&](size_t innerBlock) {
+                const size_t innerStart = innerBlock * blockSize;
+                const size_t innerEnd   = innerBlock + 1 == nInnerBlocks ? nTrain : innerStart + blockSize;
+                const size_t innerSize  = innerEnd - innerStart;
+
+                daal::services::internal::TArray<FPType, cpu> tmpArr(outerSize * innerSize);
+                FPType * tmp = tmpArr.get();
+
+                s |= euclDist.computeBatch(outerStart, outerSize, innerStart, innerSize, tmp);
+
+                PRAGMA_IVDEP
+                PRAGMA_VECTOR_ALWAYS
+                for (size_t i = outerStart; i < outerEnd; ++i)
+                {
+                    for (size_t j = innerStart; j < innerEnd; ++j)
+                    {
+                        _distances[i * nTrain + j] = tmp[(i - outerStart) * innerSize + j - innerStart];
+                    }
+                }
+            });
+        });
+
+        daal::internal::Math<FPType, cpu> math;
+        daal::threader_for(nOuterBlocks, nOuterBlocks, [&](size_t outerBlock) {
+            const size_t outerStart = outerBlock * blockSize;
+            const size_t outerEnd   = outerBlock + 1 == nOuterBlocks ? nTest : outerStart + blockSize;
+            const size_t outerSize  = outerEnd - outerStart;
+
+            daal::services::internal::TArray<FPType, cpu> tmpArr(outerSize * nTrain);
+            FPType * tmp = tmpArr.get();
+            math.vSqrt(outerSize * nTrain, _distances + outerStart * nTrain, tmp);
```

Comment:
Better to move this into Euclidean metric.
You can add some flag into the metric, do you need to compute sqrt or not.

I don't think that it will affect performance significantly

Reply:
Sqrt computing flag is added to metric